File: LeetCode_Python/permutation_sequence.py
class Solution(object):
    def factorial(self, number):
        total = 1
        for i in range(number, 0, -1): total *= i 
        return total
    
    def getPermutation(self, n, k):
        numbers = list(range(1, n + 1))
        result = []

        for _ in range(n):
            num = self.factorial(len(numbers) - 1)
            index = (k - 1) // num
            result.append(numbers.pop(index))
            k = k - index * num

        return ''.join(map(str, result))
n = 3
k = 3
sol = Solution()
ans = sol.getPermutation(n, k)
print(ans)

File: LeetCode_C/permutation_sequence.c
#include <stdlib.h>
#include <stdio.h>

int factorial(int number) {
    int total = 1;
    for (int i = number; i >= 2; i--) total *= i;
    return total;
}

char* getPermutation(int n, int k) {
    int numbers[n];
    int numbersSize = n;
    for (int i = 0; i < n; i++) numbers[i] = i + 1;
    char* ans = malloc(n + 1);

    for (int i = 0; i < n; i++) {
        int num = factorial(numbersSize - 1);
        int index = (k - 1) / num;
        ans[i] = numbers[index] + '0';
        for (int j = index; j < numbersSize - 1; j ++) numbers[j] = numbers[j + 1];
        numbersSize--;
        k = k - index * num;
    }

    ans[n] = '\0';
    return ans;
}

int main() {
    int n = 3;
    int k = 3;
    char* ans = getPermutation(n, k);
    printf("%s", ans);
    free(ans);
}